File: Assets/Loenn/triggers/InvokeMethodTrigger.lua
local trigger = {}

trigger.name = "BalintHelper/InvokeMethodTrigger"

trigger.placements = {
	{
		name = "main",
		data = {
			width = 16,
			height = 16,
			methodName = "",
			onlyOnce = true,
			argumentMode = "None",
		}
	}
}

local argumentModes = { "None", "Position", "Bounds" }

trigger.fieldInformation = {
	methodName = {
		fieldType = "string",
		description = "The name of the dynamic method (defined via a DefineMethodTrigger) to invoke when the player enters this trigger."
	},
	onlyOnce = {
		fieldType = "boolean",
		description = "Whether the method should only be invoked the first time the player enters this trigger."
	},
	argumentMode = {
		options = argumentModes,
		editable = false,
		description = "What (if anything) to pass to the dynamic method as its single argument: None, the trigger's Position (Vector2), or its Bounds (Rectangle)."
	}
}

trigger.fieldOrder = {
	"x", "y", "width", "height", "methodName", "onlyOnce", "argumentMode"
}

return trigger

File: Source/Triggers/Dynamic/InvokeMethodTrigger.cs
using Celeste.Mod.BalintHelper.Entities.Dynamic;
using Celeste.Mod.Entities;
using Microsoft.Xna.Framework;
using System;

namespace Celeste.Mod.BalintHelper.Triggers.Dynamic
{
    [CustomEntity("BalintHelper/InvokeMethodTrigger")]
    public class InvokeMethodTrigger : Trigger
    {
        public enum ArgumentMode
        {
            None,
            Position,
            Bounds
        }

        private readonly string methodName;
        private readonly bool onlyOnce;
        private readonly ArgumentMode argumentMode;
        private bool invoked;

        public InvokeMethodTrigger(EntityData data, Vector2 offset) : base(data, offset)
        {
            methodName = data.String("methodName") ?? throw new ArgumentException("no method name was provided", nameof(data));
            onlyOnce = data.Bool("onlyOnce", true);
            argumentMode = data.Enum("argumentMode", ArgumentMode.None);
        }

        public override void OnEnter(Player player)
        {
            base.OnEnter(player);

            if (onlyOnce && invoked)
            {
                return;
            }
            invoked = true;

            object?[]? args = argumentMode switch
            {
                ArgumentMode.Position => [Position],
                ArgumentMode.Bounds => [new Rectangle((int)X, (int)Y, (int)Width, (int)Height)],
                _ => null
            };

            var controller = DynamicMethodController.GetOrCreate(Scene);
            controller.Interpreter.InvokeDynamicMethod(methodName, out _, args);
        }
    }
}
